File: GLSL/3DPrimitivesDrawing/source/ShaderProgram/ShadersManager.cpp
#include "ShadersManager.h"

ShadersManager* ShadersManager::_ref = NULL;

ShadersManager::ShadersManager()
{

}

ShadersManager::~ShadersManager()
{

}

ShadersManager* ShadersManager::GetInstance()
{
	if (_ref == NULL)
		_ref = new ShadersManager();

	return _ref;
}



ShaderProgram* ShadersManager::CreateShaderProgram(string vertexShader, string fragmentShader)
{
	string key = vertexShader;
	key.append(":").append(fragmentShader);

	ShaderProgram* shaderProgram = NULL;

	if (shadersMap.find(key) != shadersMap.end())
	{
		ShaderInfo* shaderInfo = shadersMap.find(key)->second;
		shaderInfo->IncreaseRefCount();
		shaderProgram = shaderInfo->GetShaderProgram();
	}
	else
	{
		shaderProgram = new ShaderProgram(vertexShader.c_str(), fragmentShader.c_str());
		ShaderInfo* shaderInfo = new ShaderInfo(shaderProgram);
		shadersMap[key] = shaderInfo;
	}

	return shaderProgram;
}

void ShadersManager::DeleteShaderProgram(ShaderProgram* shaderProgram)
{
	string vertexShader = shaderProgram->GetVertexShaderFilePath();
	string fragmentShader = shaderProgram->GetFragmentShaderFilePath();

	string key = vertexShader.append(":").append(fragmentShader);

	if (shadersMap.find(key) != shadersMap.end())
	{
		ShaderInfo* shaderInfo = shadersMap.find(key)->second;
		shaderInfo->DecreaseRefCount();

		if (shaderInfo->GetRefCount() == 0)
		{
			delete shaderInfo->GetShaderProgram();
			delete shaderInfo;
			shadersMap.erase(shadersMap.find(key));
		}
	}
}

void ShadersManager::DeleteInstance()
{
	if (_ref)
	{
		delete _ref;
		_ref = NULL;
	}
}

File: WebGL/NodeJS/public/src/ShadersManager.js
var shadersManagerInstanceCount = 0;

class ShaderInfo
{
	constructor(shaderProgram)
	{
		this._shaderProgram = shaderProgram;
		this._refCount = 1;
	}
	
	getRefCount()
	{
		return this._refCount;
	}

	increaseRefCount()
	{
		this._refCount++;
	}

	decreaseRefCount()
	{
		this._refCount--;

		if (this._refCount < 0)
		{
			throw "Reference count should not be negative number";
		}
	}

	getShaderProgram()
	{
		return this._shaderProgram;
	}
}

	
class ShadersManager
{
	constructor()
	{
		shadersManagerInstanceCount++;
		
		if(shadersManagerInstanceCount > 1)
		{
			throw "More than one object creation is not allowed for ShadersManager Class";
		}
		
		this._shadersMap = new Map(); 
		
	}
	
	async createShaderProgram(vertexShader, fragmentShader)
	{
		var key = vertexShader+":"+fragmentShader;

		var shaderProgram = null;

		if (this._shadersMap.has(key))
		{
			console.log("Shader Manager already contains this shader");
			var shaderInfo = this._shadersMap.get(key);
			shaderInfo.increaseRefCount();
			shaderProgram = shaderInfo.getShaderProgram();			
		}
		else
		{
			console.log("Shader Manager doesn't contains this shader");
			var shaderProgram = new ShaderProgram();
			await shaderProgram.init(vertexShader, fragmentShader);
			var shaderInfo = new ShaderInfo(shaderProgram);
			this._shadersMap.set(key, shaderInfo);
		}

		return shaderProgram;
	}
	
	deleteShaderProgram(shaderProgram)
	{
		var vertexShaderPath = shaderProgram.getVertexShaderFilePath();
		var fragmentShaderPath = shaderProgram.getFragmentShaderFilePath();

		var key = vertexShaderPath+":"+fragmentShaderPath;

		if (this._shadersMap.has(key))
		{
			var shaderInfo = this._shadersMap.get(key);
			shaderInfo.decreaseRefCount();

			if (shaderInfo.getRefCount() == 0)
			{
				shaderInfo.getShaderProgram().delete();
				this._shadersMap.delete(key);
			}
		}
	}	
}


var shadersManager = new ShadersManager();
